### ISSUE TITLE
The inferred type of "X" cannot be named without a reference to "Y". This is likely not portable.

### ISSUE THREAD

**Issue by GuillaumeLagrange** [swarmion/swarmion#508]
With how PNPM manages packages, in a swarmion project we can easily run into the following issue:

- Package `@my-app/shared-foo` packages has `package-a` as dependency, exporting a `TypeA` type
    - In this package, define a function `sharedFunction` whose return type depends directly on `TypeA`
- Service `@my-app/backend-user` has `@my-app/shared-foo` as dependency and uses `sharedFunction`.

In `@my-app/backend-user`:
```typescript
const bar = sharedFunction();
```

This will create an LSP warning 
`
The inferred type of 'bar' cannot be named without a reference to '../path/to/shared-foo/node_modules/package-a'. This is likely not portable. A type annotation is necessary.
`

Minimal reproduction on the full stack example: 8442a7b3353b7559a16da83fbf364daf124bd8cc (line 15@`examples/swarmion-full-stack/frontend/app/src/main.tsx`)

**Comment by GuillaumeLagrange** [comment]
After investigation, I have found a less imperfect fix:

In `@my-app/shared-foo`, reexport `TypeA` from `package-a`

It suppresses warnings, but is imperfect because there is no way to know there is an issue until you encounter it by consuming your package

**Comment by adriencaccia** [comment]
Thanks for the issue @GuillaumeLagrange, you did a great job pinpointing the issue to this simple cause 💪 

> * Investigate why the issue happens only in the LSP and node when running test-type

Here it would be better to put: _... and not when running `tsc` directly in the cli through the `test-type` script_

Can you open an issue on the `pnpm` repo and link this issue? I think this is the best way forward, hopefully someone there would be able to help us 🙂

In the meantime, the fix you mentioned can be used. And I would recommend adding a comment linking to this issue above the `export type ...` statement.

**Comment by fargito** [comment]
@GuillaumeLagrange is this fixed by #541 ?

**Comment by GuillaumeLagrange** [comment]
Nope, I have managed to reproduce the issue with tsc, and opened an issue on pnpm repo to see if anybody has an idea of what is happening

**Comment by GuillaumeLagrange** [comment]
An answer on the pnpm issue pointed out a decent workaround that will do until the root cause is fixed:

![image](https://user-images.githubusercontent.com/19265358/219581275-e119d294-391e-437f-9ea1-1a335f96cc85.png)

**Comment by fargito** [comment]
This seems sensible, we'll keep the issue open during this time

**Comment by adriencaccia** [comment]
I just had the same problem on a project, but this time on a _service_ package, not a shared code library.
Since this project is not aimed to be shared, I could solve the error by removing all the properties related to `declaration`: 

![image](https://user-images.githubusercontent.com/19605940/223144141-37e6a083-4370-40a6-bc4d-802a56ddffec.png)

It seems that this issue is happening more and more on more and more monorepos with `pnpm` and `TypeScript`.

Hopefully, this fix can help when the above fix mentioned by @GuillaumeLagrange does not work.

**Comment by adriencaccia** [comment]
@GuillaumeLagrange we ought to try this https://github.com/microsoft/TypeScript/issues/42873#issuecomment-1464721715

**Comment by pkerschbaum** [comment]
FYI I added [a comment in TypeScript#47663](https://github.com/microsoft/TypeScript/issues/47663#issuecomment-1519138189) which has workarounds for this issue when using pnpm monorepos.

**Comment by GuillaumeLagrange** [comment]
Thank you for this, it really sums up quite clearly the different workarounds and their eventual downsides.

Not much more we can do than wait on a root fix from either Typescript or pnpm now, not sure how they would go about that though.

**Comment by adriencaccia** [comment]
Thanks @pkerschbaum for the link to the detailed explanation!

Since the problem occurs only on packages not meant to be consumed by others, we will go with the solution of setting `declaration: false`.
